File: src/go/dsa/algorithms.go
package dsa

import "cmp"

func BinarySearchList[CT cmp.Ordered](list *MyList[CT], key CT) int64 {
	return -1
}

func BinarySearchListIndices[CT cmp.Ordered](list *MyList[CT],
	startIndex uint32, endIndex uint32, key CT) int64 {
	return -1
}

func BinarySearchArray[CT cmp.Ordered](array []CT, key CT) int64 {
	return -1
}

func BinarySearchArrayIndices[CT cmp.Ordered](array []CT, startIndex uint32,
	endIndex uint32, key CT) int64 {
	return -1
}

func MergesortList[CT cmp.Ordered](list *MyList[CT]) {

}

func MergesortArray[CT cmp.Ordered](array []CT) {

}

func QuicksortList[CT cmp.Ordered](list *MyList[CT]) {

}

func QuicksortArray[CT cmp.Ordered](array []CT) {

}

func HeapsortList[CT cmp.Ordered](list *MyList[CT]) {

}

func HeapsortArray[CT cmp.Ordered](array []CT) {

}

File: src/rust/dsa/src/algorithms.rs
use crate::mylist::MyList;

// In C++ std::span(T[]) is the equivalent of &[T]

pub fn binary_search_list<T: PartialOrd>(_list: &impl MyList<T>, _key: T) -> i64 {
    -1
}

pub fn binary_search_list_indices<T: PartialOrd>(
    _list: &impl MyList<T>,
    _start_index: u32,
    _end_index: u32,
    _key: T,
) -> i64 {
    -1
}

pub fn binary_search_array<T: PartialOrd>(_list: &[T], _key: T) -> i64 {
    -1
}

pub fn binary_search_array_indices<T: PartialOrd>(
    _list: &[T],
    _start_index: u32,
    _end_index: u32,
    _key: T,
) -> i64 {
    -1
}

pub fn mergesort_list<T: PartialOrd>(_list: &mut impl MyList<T>) {}

pub fn mergesort_array<T: PartialOrd>(_list: &mut [T]) {}

pub fn quicksort_list<T: PartialOrd>(_list: &mut impl MyList<T>) {}

pub fn quicksort_array<T: PartialOrd>(_list: &mut [T]) {}

pub fn heapsort_list<T: PartialOrd>(_list: &mut impl MyList<T>) {}

pub fn heapsort_array<T: PartialOrd>(_list: &mut [T]) {}
